proposed/2021/LicenseInspection.md

```diff
@@ -0,0 +1,133 @@
+- Start Date (2021-05-11)
+- GitHub Issue N/A
+- GitHub PR TBC
+
+## Summary
+
+This command will allow someone to inspect all the licenses that are specified for the referenced packages, as well as transient packages, of a project.
+
+## Motivation
+
+The motivation is to give people more insights into the licenses that are used in a project, ensuring they can stay on top of compliancy requirements.
+
+## Explanation
+
+On the new project you're building, there are dependencies on NuGet packages from a remote feed, and those packages in turn have their own dependencies. This has resulted in an opaque view of what licenses are being consumed by the project.
+
+Your companies legal department has outlined a list of open source licenses that it is comfortable consuming and others that it wants to avoid using.
+
+To avoid accidentally consuming an unsupported license we need a way to view the list and ideally fail builds when we use one of these licenses.
+
+### Functional explanation
+
+#### Local scenario
+
+You have just added a new NuGet dependency into the project and want to ensure it doesn't bring in any unsupported licenses.
+
+From the command line you run the license command:
+
+```bash
+$> dotnet nuget license
+```
+
+This outputs a human readable view of the license information:
+
+```
+Project dotnet-delice
+License Expression: MIT
+├── There are 10 occurances of MIT
+├─┬ Conformance:
+│ ├── Is OSI Approved: true
+│ ├── Is FSF Free/Libre: true
+│ └── Included deprecated IDs: false
+└─┬ Packages:
+  ├── FSharp.Core
+  ├── Microsoft.NETCore.App
+  ├── Microsoft.NETCore.DotNetAppHost
+  ├── Microsoft.NETCore.DotNetHostPolicy
+  ├── Microsoft.NETCore.DotNetHostResolver
+  ├── Microsoft.NETCore.Platforms
+  ├── Microsoft.NETCore.Targets
+  ├── NETStandard.Library
+  ├── Newtonsoft.Json
+  └── System.ComponentModel.Annotations
+```
+
+Satisfied with the licenses, you continue on with the task at hand.
+
+#### Blocking builds
+
+It's time to update your build pipeline to avoid releasing anything changes that introduce unsupported licenses. A new step is added to the pipeline YAML:
+
+```yml
+- name: Validate Licenses
+  runs: |
+    LICENSES=$(dotnet nuget license --json | jq '<query for allowed licenses>')
+    if ($LICENSES) then
+        exit 1
+    fi
+```
+
+This (pseduocode version) will dump the licenses to a JSON structure that can be parsed with a tool such as [jq](https://stedolan.github.io/jq/) and inspected for invalid licenses. If any are found an error code is returned, stopping the build from completing successfully.
+
+### Technical explanation
+
+The first technical challenge for this is the inconsistent nature of which licenses are provided by NuGet packages. While the [`licenseUrl` field was deprecated](https://github.com/NuGet/Announcements/issues/32), some projects haven't adopted the new format (or older packages that predate the deprecation are in use), making it difficult to determine what the license of a project is.
```

Comment:
- I would suggest computing similarity as a command option. Making a compliance decision without reading the license can be risky, as just one word can change the whole license meaning.
- How about scope-spliting this spec into the following scenarios?
  1. first, listing licenses of all packages, e.g. `dotnet list package --show-license`, just showing license types
  1. then, making 'similarity aggregation' in SDPX licenses, as an option
  1. later, making 'similarity aggregation' license files, with similarity selection
  1. finally, licenseUrl, as an option

I cannot make any guarantees of whether or not this going to be implemented, but, implementing the first two scenarios looks feasible. We are more than happy to review a community PR :)

Reply:
> * I would suggest computing similarity as a command option. Making a compliance decision without reading the license can be risky, as just one word can change the whole license meaning.

This wouldn't make compliance decisions, it's merely surfacing the information that is required to allow decision makers to make said decision. A level of similarity is required, which is why something such as Sørensen–Dice coefficient is used to determine the license (if it's not explicitly set), otherwise you'd be provided with what is mostly unhelpful information (from when I did the initial proposal, the usage of the SPDX ID's for licensing was low relative to a dedicated URL).

> * How about scope-spliting this spec into the following scenarios?
>   
>   1. first, listing licenses of all packages, e.g. `dotnet list package --show-license`, just showing license types
>   2. then, making 'similarity aggregation' in SDPX licenses, as an option
>   3. later, making 'similarity aggregation' license files, with similarity selection
>   4. finally, licenseUrl, as an option
> 
> I cannot make any guarantees of whether or not this going to be implemented, but, implementing the first two scenarios looks feasible. We are more than happy to review a community PR :)

Dropping license similarity based on file content I feel would really decrease the value of using this, as my past testing has indicated that the usage of SPDX identifiers in NuGet packages for license indication is relatively low, and that's ultimately why I added the feature to `dotnet-delice` to do template comparisons.

Given that this has already been implemented once, I'm probably overestimating the simplicity of implementing it again (converting the F# to C#), but doing so with some additional branches that allows you to opt-out of SPDX template comparisons wouldn't be a huge overhead in the process.